components/script/dom/gamepadhapticactuator.rs

```diff
@@ -0,0 +1,381 @@
+
+        if self.playing_effect_promise.borrow().is_some() {
+            let trusted_promise = TrustedPromise::new(
+                self.playing_effect_promise
+                    .borrow()
+                    .clone()
+                    .expect("Promise is null!"),
+            );
+            *self.playing_effect_promise.borrow_mut() = None;
+            let _ = self.global().gamepad_task_source().queue(
+                task!(preempt_promise: move || {
+                    let promise = trusted_promise.root();
+                    let message = DOMString::from("preempted");
+                    promise.resolve_native(&message);
+                }),
+                &self.global(),
+            );
+        }
+
+        if !self.effects.contains(&type_) {
+            playing_effect_promise.reject_error(Error::NotSupported);
+        }
+
+        assert!(self.playing_effect_promise.borrow().is_none());
+
+        *self.playing_effect_promise.borrow_mut() = Some(playing_effect_promise.clone());
+        self.effect_sequence_id.set(self.sequence_id.get());
+
+        // XXX The spec says we SHOULD also pass a playEffectTimestamp for more precise playback timing
+        // when start_delay is non-zero, but this is left more as a footnote without much elaboration.
+
+        let params = DualRumbleEffectParams {
+            duration: params.duration as f64,
+            start_delay: params.startDelay as f64,
+            strong_magnitude: *params.strongMagnitude,
+            weak_magnitude: *params.weakMagnitude,
+        };
+        let event = EmbedderMsg::PlayGamepadHapticEffect(
+            self.gamepad_index as usize,
+            embedder_traits::GamepadHapticEffectType::DualRumble(params),
+        );
+        self.global().as_window().send_to_embedder(event);
+
+        playing_effect_promise
+    }
+
+    /// <https://www.w3.org/TR/gamepad/#dom-gamepadhapticactuator-reset>
+    fn Reset(&self) -> Rc<Promise> {
+        let in_realm_proof = AlreadyInRealm::assert();
+        let promise = Promise::new_in_current_realm(InRealm::Already(&in_realm_proof));
+
+        let document = self.global().as_window().Document();
+        if !document.is_fully_active() {
+            promise.reject_error(Error::InvalidState);
+            return promise;
+        }
+
+        self.sequence_id.set(self.sequence_id.get().wrapping_add(1));
+
+        if self.playing_effect_promise.borrow().is_some() {
+            let trusted_promise = TrustedPromise::new(
+                self.playing_effect_promise
+                    .borrow()
+                    .clone()
+                    .expect("Promise is null!"),
+            );
+            *self.playing_effect_promise.borrow_mut() = None;
+            let _ = self.global().gamepad_task_source().queue(
+                task!(preempt_promise: move || {
+                    let promise = trusted_promise.root();
+                    let message = DOMString::from("preempted");
```

Comment:
this part belongs in the handler of `HapticEffectStopListener`, and I think this should also include the other promise and resolve it with "completed"(was this resolved in the spec?)

Reply:
In the PR spec the preempted promise handling isn't done in parallel anymore, just the completed promise steps along with the haptic effect stop